### DISCUSSION TITLE
Jacobians of materials with coupled variables

### DISCUSSION THREAD

**Discussion by NateAM** [idaholab/moose#25796] [Q&A General] [answered]
I have a material that is a function of various solution variables (`v`). I have been able to compute the Jacobians within the material property and expose the Jacobians to a kernel. This kernel however does not explicitly depend on `u` so when `computeOffDiagJacobian` is called in my kernel the Kernel doesn't update.

Because these are very general materials, I don't always know what they will be functions of. Is there a way to have my kernel add the coupled solution variables for a given material property to it's own without having to add that to the input definition?

**Comment by GiudGiud** [comment]
Hello

Computing the Jacobian inside a material is not standard. It s the kernel's role to compute the Jacobian.

Did you verify that `computeOffDiagJacobian` is called? With some preconditioning in PFJNK it is not called unless a parameter is passed. 

A kernel will not be able to modify the way a material property is being computed. The material property is pre-computed for all quadrature points when considering an element in the residual loop & jacobian loop

Guillaume

**Reply by NateAM** [reply]
> Computing the Jacobian inside a material is not standard. It s the kernel's role to compute the Jacobian.

Sometimes you have materials that are functions of solution variables though (solid mechanics has this situation). Computing the consistent tangents of those materials in the same material object would seem seem to make sense, at least to me. Maybe I'm not thinking about it the way that the MOOSE dev team does.

> Did you verify that computeOffDiagJacobian is called? With some preconditioning in PFJNK it is not called unless a parameter is passed.

Yes, I have gotten it to call. A solution is just to add a call to `addCoupledVar` to the kernel that the user has to specify which variables were used when computing the material. That works, but I was hoping there would be something less prone to error. 

> A kernel will not be able to modify the way a material property is being computed. The material property is pre-computed for all quadrature points when considering an element in the residual loop & jacobian loop

I don't want the kernel to modify what the material is doing, I want a kernel that consumes a material to know that it has more dependencies that will need Jacobian contributions due to the material's dependencies.

**Reply by GiudGiud** [reply, marked as answer]
If you use AD, it will capture that additional dependencies naturally when computeResidual is called for computing the Jacobian. 

If you do not use AD, then implementing the computeOffDiagJacobian is where you need to capture all these dependencies, manually. You ll need to add the derivatives of the material properties with regards to the variables coupled in

**Reply by NateAM** [reply]
> If you use AD, it will capture that additional dependencies naturally when computeResidual is called for computing the Jacobian.

Okay, that makes sense. This is a personal preference, but I like to use hand-coded Jacobians as a form of verification. I do understand the appeal of AD though.

> If you do not use AD, then implementing the computeOffDiagJacobian is where you need to capture all these dependencies, manually. You ll need to add the derivatives of the material properties with regards to the variables coupled in

Okay, that's disappointing but I understand. Thanks!

**Reply by lindsayad** [reply]
@NateAM i think it makes sense to do the derivative calculations in the material especially if you are going to use it in multiple residual objects. But as you’ve noted there is no way to get around the requirement (at least not currently) that you need the explicit coupling via addCoupledVar in order to get the off diagonal Jacobian call for that variable. Sorry for the inconvenience!

**Reply by NateAM** [reply]
> @NateAM i think it makes sense to do the derivative calculations in the material especially if you are going to use it in multiple residual objects. But as you’ve noted there is no way to get around the requirement (at least not currently) that you need the explicit coupling via addCoupledVar in order to get the off diagonal Jacobian call for that variable. Sorry for the inconvenience!

No worries! I just get so used to most things being possible here that I'm surprised when I run into something that isn't available yet!

**Reply by GiudGiud** [reply]
Thinking about it during my run, this seems totally fixable. I ll raise an issue. No timeline on the fix though

**Reply by NateAM** [reply]
That would be great @GiudGiud! I have a workaround that is fine for the present but it would be nice if there were more guardrails to keep me from messing up an input deck.

**Reply by joshuahansel** [reply]
I wasn't sure if this was still of interest since it's been so long. But to hand-code a Jacobian in a kernel, you need to know the `jvar` of the coupled variable (which you get from `coupled("somevar")`), so I don't know what you were doing in `computeOffDiagJacobian` (where were you getting the `jvar` you were computing the Jacobian for, if the variable wasn't coupled to the Kernel?). Obviously there's nothing we can do to prevent users from forgetting to code off-diagonal Jacobian contributions from materials, but at the least maybe we can throw an error/warning if a material has dependencies on variables that are not coupled into the kernel.

**Reply by NateAM** [reply]
> I wasn't sure if this was still of interest since it's been so long. But to hand-code a Jacobian in a kernel, you need to know the `jvar` of the coupled variable (which you get from `coupled("somevar")`), so I don't know what you were doing in `computeOffDiagJacobian` (where were you getting the `jvar` you were computing the Jacobian for, if the variable wasn't coupled to the Kernel?). Obviously there's nothing we can do to prevent users from forgetting to code off-diagonal Jacobian contributions from materials, but at the least maybe we can throw an error/warning if a material has dependencies on variables that are not coupled into the kernel.

@joshuahansel thanks for the comment here. I don't know if @GiudGiud ended up adding material variable dependencies to the kernel or if I implemented a workaround (it's been quite a while since I touched this part of my code) but at least my application seems to be working correctly at this point. If a kernel's variable dependencies hasn't already been set to update if a material it requires depends on something a warning would be really useful to keep from input deck mistakes.

**Reply by YaqiWang** [reply]
You could also let the material evaluate more material properties that store the derivative of the material property with respect  solution. Then in the kernel you can obtain these more material properties and use them in `computeOffDiagJacobian`.

**Reply by NateAM** [reply]
> You could also let the material evaluate more material properties that store the derivative of the material property with respect solution. Then in the kernel you can obtain these more material properties and use them in `computeOffDiagJacobian`.

Hi @YaqiWang yes, that's what I was doing. The issue I was having at the time was that my material might depend on things that my kernel doesn't explicitly and I wanted the kernel to detect the variables linked to the material and propagate that to the kernel. I've been able to get it working but I was hoping for something more automatic.
